File: Daily Resources/Day11/DSA - Tree Solutions/distance_queries.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 2e5 + 5, LOG = 20;
vector<int> tree[N];
int up[N][LOG], depth[N];

void dfs(int u, int p) {
    up[u][0] = p;
    for (int i = 1; i < LOG; ++i)
        if (up[u][i-1] != -1)
            up[u][i] = up[up[u][i-1]][i-1];
    for (int v : tree[u]) {
        if (v == p) continue;
        depth[v] = depth[u] + 1;
        dfs(v, u);
    }
}

int lca(int a, int b) {
    if (depth[a] < depth[b]) swap(a, b);
    for (int i = LOG - 1; i >= 0; --i)
        if (depth[a] - (1 << i) >= depth[b])
            a = up[a][i];
    if (a == b) return a;
    for (int i = LOG - 1; i >= 0; --i)
        if (up[a][i] != -1 && up[a][i] != up[b][i])
            a = up[a][i], b = up[b][i];
    return up[a][0];
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n, q;
    cin >> n >> q;
    for (int i = 0; i < n - 1; ++i) {
        int a, b; cin >> a >> b;
        tree[a].push_back(b);
        tree[b].push_back(a);
    }

    dfs(1, -1);

    while (q--) {
        int a, b;
        cin >> a >> b;
        int l = lca(a, b);
        cout << depth[a] + depth[b] - 2 * depth[l] << '\n';
    }
}

File: Daily Resources/Day11/DSA - Tree Solutions/distance_queries.py
import sys
input = sys.stdin.read
sys.setrecursionlimit(1 << 25)

data = input().split()
n, q = int(data[0]), int(data[1])
edges = data[2:2 + 2 * (n - 1)]
queries = data[2 + 2 * (n - 1):]

LOG = 20
tree = [[] for _ in range(n + 1)]
up = [[-1]*LOG for _ in range(n + 1)]
depth = [0]*(n + 1)

for i in range(0, len(edges), 2):
    a, b = int(edges[i]), int(edges[i+1])
    tree[a].append(b)
    tree[b].append(a)

def dfs(u, p):
    up[u][0] = p
    for i in range(1, LOG):
        if up[u][i-1] != -1:
            up[u][i] = up[up[u][i-1]][i-1]
    for v in tree[u]:
        if v != p:
            depth[v] = depth[u] + 1
            dfs(v, u)

dfs(1, -1)

def lca(a, b):
    if depth[a] < depth[b]:
        a, b = b, a
    for i in range(LOG - 1, -1, -1):
        if depth[a] - (1 << i) >= depth[b]:
            a = up[a][i]
    if a == b:
        return a
    for i in range(LOG - 1, -1, -1):
        if up[a][i] != -1 and up[a][i] != up[b][i]:
            a = up[a][i]
            b = up[b][i]
    return up[a][0]

out = []
for i in range(0, len(queries), 2):
    a, b = int(queries[i]), int(queries[i+1])
    l = lca(a, b)
    out.append(str(depth[a] + depth[b] - 2 * depth[l]))

print('\n'.join(out))
